Add documentation for state windows (#274)
tremor-rs/tremor-www: docs/language/pipelines.md

- `interval`: Time interval in nanoseconds after which the window closes.

##### Windows Based on State

> The most flexible windows are code or state based windows. Instead of being fixed on the timestamp

Reviewer: ```suggestion
The most flexible windows are code- or state-based windows. Instead of being fixed on the timestamp
```

> or a message cout it allows to express the window logic of closing and emitting in tremor-script.

Reviewer: ```suggestion
or a message count, it allows to express the window logic of closing and emitting in tremor-script.
```

Reviewer: We need to explain the mechanics of `closing` and `emitting`. I think we do not do it anywhere else, and tremor users cannot be expected to understand how this works.

Author: Good call on this, I'll rewrite it to omit special terms and make it more user friendly to read

This requires two parts:

1) the `state` section of the window defines the inital state, this is the same as `state` for the script operator.

> 2) the `script from tick` section defines the logic that is executed on the periodic ticks to ensure windows that don't get new events can be closed and emitted.

Reviewer: ```suggestion
2) the `script from tick` section defines the logic that is executed on periodic ticks to ensure windows that don't get new events can be closed and emitted.
```

We need to document what the tick interval is. Is it still 100ms?

Reviewer: We need to add that the normal `script` is executed for each incoming event, and there `event` and `$` are available, no?

Reviewer: We also need to document that for ticks, the current tick time can be fetched by using `tremor::system::ingest_ns()`.

Author: ja good call!

> The result of the script and tick script can emit a `array` of two elements. The first element defines if the widow should emit. The second one if the value should be included in this window (before emitting) or included in the ne

Reviewer: ```suggestion
The result of the script and tick script can emit either a boolean value or an `array` of two elements. The first element defines if the widow should emit its accumulated data. The second one defines if the current value should be included in this window (before emitting) or included in the next emit.
```

end;
```

> the same as state window would read:

Reviewer: ```suggestion
the above window as state window would be implemented as follows:
```

```trickle
use std::time::nanos;
use tremor::system;

define window fifteen_secs from tumbling
state
  null
script
  match state of
      null =>

>         let state = system::nanotime();

Reviewer: this one should use `tremor::system::ingest_ns()`, no? And for making this example showing all the possible constructs of a state based window it would be cool to add a `script from tick` section too. (This would also make it behave exactly as the tumbling window on time)

Author: yes, you're right! good call :)
